[don't merge without prior discussion] fix: don't sort author lists, except for segment grouping purposes

corneliusroemer opened the pull request:
resolves #2650

preview URL: https://author-order.loculus.org

### Summary

We started sorting author lists when it turned out that different segments of the same isolate could have different author orders. We should have done the sorting only for grouping purposes and not when submitting to Loculus.

This PR fixes that by only sorting for grouping purposes.

Note: this will cause ingest to submit revisions for existing deployments as the ingest hash changes.

### Screenshot

corneliusroemer commented on ingest/scripts/call_loculus.py:
```diff
@@ -483,7 +483,7 @@ def record_factory(*args, **kwargs):
         logger.info("Getting submitted sequences")
         response = get_submitted(config)
-        Path(output).write_text(json.dumps(response), encoding="utf-8")
+        Path(output).write_text(json.dumps(response, indent=4, sort_keys=True), encoding="utf-8")
```
This change is unrelated - could split it into separate PR

Just to make output diffs more stable

corneliusroemer replied:
```suggestion
        Path(output).write_text(json.dumps(response, indent=4), encoding="utf-8")
```

corneliusroemer commented on ingest/scripts/group_segments.py:
```diff
@@ -94,27 +108,36 @@ def main(
     equivalence_classes: EquivalenceClasses = defaultdict(lambda: defaultdict(list))
     for accession, values in segment_metadata.items():
-        group_key = tuple((field, values[field]) for field in shared_fields if values[field])
+        # Author order sometimes varies despite segment apparently the same assembly
```
```suggestion
        # Author order sometimes among segments from same isolate
```

corneliusroemer commented on ingest/scripts/group_segments.py:
```diff
@@ -48,6 +61,7 @@ class Config:
     segmented: bool
 
 
+# TODO: Better name needed, what is special about these fields?
```
segment is by definition different across segments, so this is unnecessary to make configurable (hard coded special field)

corneliusroemer commented on ingest/scripts/prepare_metadata.py:
```diff
@@ -145,7 +145,7 @@ def main(
     meta_dict = {rec[fasta_id_field]: rec for rec in metadata}
 
-    Path(output).write_text(json.dumps(meta_dict, indent=4), encoding="utf-8")
+    Path(output).write_text(json.dumps(meta_dict, indent=4, sort_keys=True), encoding="utf-8")
```
```suggestion
    Path(output).write_text(json.dumps(meta_dict, indent=4), encoding="utf-8")
```

theosanderson commented on ingest/scripts/group_segments.py:
```diff
@@ -94,27 +108,38 @@ def main(
-    first_row = next(iter(segment_metadata.values()))
-    if not first_row:
+    # Dynamically determine the fields that are present in the metadata
+    some_row = next(iter(segment_metadata.values()))
```
Is this _not_ the first row? If it is still that then I find the original version clearer fwiw (though I found your changes above very helpful)

corneliusroemer replied:
Since we're using Python >=3.7 there _is_ a first element, so you're right!

https://mail.python.org/pipermail/python-dev/2017-December/151283.html

theosanderson commented on ingest/scripts/group_segments.py:
```diff
@@ -94,27 +108,38 @@ def main(
     equivalence_classes: EquivalenceClasses = defaultdict(lambda: defaultdict(list))
     for accession, values in segment_metadata.items():
-        group_key = tuple((field, values[field]) for field in shared_fields if values[field])
+        # Author order sometimes among segments from same isolate
```
Missing word?

corneliusroemer replied:
true!

theosanderson commented on ingest/scripts/group_segments.py:
```diff
@@ -189,11 +214,15 @@ def main(
+                else:
+                    msg = f"Assertion failed: values for group must be identical: {values}"
+                    raise ValueError(msg)
+            row[field] = deduplicated_values[0]
```
Just to be sure: this is sorted such that `0` will always be the same for the same data (even if the data comes up in different orders)? (I'm not doubting it, just checking)

theosanderson replied:
Yes it is :)! Just above

theosanderson reviewed (commented):
Very good work - looks good to me. I don't think we should merge this until tomorrow at the point that we're ready to start looking into bringing this to prod, because once we do we block on bringing in other new things (if something even more urgent were to come up in the intervening time).

corneliusroemer commented:
Thanks for the review @theosanderson, I've addressed your points - I won't merge then, let's coordinate tomorrow on how to phase this into production!

anna-parker reviewed (approved):
Thanks so much for doing this @corneliusroemer!

anna-parker commented:
One suggestion
```
BEGIN TRANSACTION;

SELECT accession
INTO TEMPORARY TABLE temp_accessions
FROM sequence_entries
WHERE submitter = 'insdc_ingest_user' AND version = 2;

DELETE FROM sequence_entries
WHERE accession IN (SELECT accession FROM temp_accessions) AND version = 1;

UPDATE sequence_entries
SET version = 1
WHERE submitter = 'insdc_ingest_user' AND version = 2;

DROP TABLE temp_accessions;


SELECT accession
INTO TEMPORARY TABLE temp_accessions_preprocessed
FROM sequence_entries_preprocessed_data
WHERE submitter = 'insdc_ingest_user' AND version = 2;

DELETE FROM sequence_entries_preprocessed_data
WHERE accession IN (SELECT accession FROM temp_accessions_preprocessed) AND version = 1;

UPDATE sequence_entries_preprocessed_data
SET version = 1
WHERE submitter = 'insdc_ingest_user' AND version = 2;

DROP TABLE temp_accessions_preprocessed;

COMMIT;
```

anna-parker commented:
DO $$ 
DECLARE 
    rec RECORD; 
BEGIN 
    -- Loop over each accession where submitter is 'insdc_ingest_user' and version is 2
    FOR rec IN 
        SELECT accession 
        FROM sequence_entries 
        WHERE submitter = 'insdc_ingest_user' AND version = 2 
    LOOP 
        -- Start a new transaction for each accession
        BEGIN 
            -- Delete entries where accession is found and version is 1
            DELETE FROM sequence_entries 
            WHERE accession = rec.accession AND version = 1;

            -- Update the version to 1 for the current accession and version 2
            UPDATE sequence_entries 
            SET version = 1 
            WHERE accession = rec.accession AND version = 2;

            -- Commit the transaction
            COMMIT; 
        EXCEPTION 
            -- If there's an error, rollback the transaction and continue with the next accession
            WHEN OTHERS THEN 
                ROLLBACK; 
        END; 
    END LOOP; 
END $$;

anna-parker commented:
Weird I didn't think we could have duplicate ingest anymore: 
<img width="306" alt="image" src="https://github.com/user-attachments/assets/a1bfe46f-1a8c-424f-8dba-4cdbfa21691d">

theosanderson commented:
Yes, that is mysterious

corneliusroemer commented:
Yeah it's odd, but I can't see why author order would have this effect?

anna-parker commented:
uggh... https://github.com/loculus-project/loculus/pull/2837 is now at the same state as this, I just rebased but it had the exact same bug in west nile and only in west nile ... so I do think it is this PR...

theosanderson commented:
It doesn't seem to have the duplication right now? https://ing-upd.loculus.org/ did we change sthng?

anna-parker commented:
I rebased to see if that had an impact - now https://ing-upd.loculus.org/  has a duplication in cchf

anna-parker commented:
But maybe this is more linked to our ingest deployment? Is there any chance ingest pods can run concurrently? 

I'm uncertain why this would only show up in this PR... maybe now ingest takes longer?

anna-parker commented:
Lets maybe not sort on keys? e.g. https://github.com/loculus-project/loculus/pull/2837/commits/1d833f9ac5eb94e9bbfcf7d2ef5581851f97a966 - adding the sort makes ingest slower and we want to minimize time between the get-original-metadata call and the submit calls.

corneliusroemer commented:
@anna-parker have you measured how much this changes? I'd be _very_ surprised if the sort takes more than a second. Sorting is fast, this is a small dataset.

anna-parker commented:
Yeah its not a big difference, e.g. for west nile
```
10:37:12    INFO ( prepare_metadata.py: 151) - Unsorted write took: 0.32383300000000004 
10:37:13    INFO ( prepare_metadata.py: 154) - Sorted write took: 0.356398 
```
But found it suspicious we saw this for the organisms with the most data
